### PR TITLE
Assigning task with rabbit mq

### PR DESCRIPTION
**Pull request by huyluong22**
I have created a POST API /api/user/{userId}/task-assign in the user-service so that, in the hypothetical case where role-based access control is implemented, users with the admin role can assign tasks to other users.
image
When we send this request body, the user-service will replace the userId in the request with the corresponding Id from the User table and then forward it to the task-service. The task-service will handle saving the user and assigned task information into the assignment table, then send a response back to the user-service via RabbitMQ.
image
image
After the message is sent to RabbitMQ, it will remain in the Queued messages. At this point, the consumer in the user-service will retrieve the information from the Queued messages, create a notification, and save it to the corresponding table associated with the user
image

image

### CODE REVIEW

**Comment by thaibinh156** [comment]
LGTM

File: task-service/src/main/java/com/infodation/task_service/controllers/TasksController.java [hunk 1]
```diff
@@ -1,27 +1,73 @@
...
 
-    public TasksController(ITaskService taskService) {
+    public TasksController(ITaskService taskService, RabbitTemplate rabbitTemplate) {
         this.taskService = taskService;
+        this.rabbitTemplate = rabbitTemplate;
+    }
+    public void sendNotificationToUser(TaskAssignmentDTO message) {
+        rabbitTemplate.convertAndSend("sendNotification", message);
+        log.info("Messages sent: -----> " + message);
+    }
```
Comment (SonNgocTruong) [inline task-service/src/main/java/com/infodation/task_service/controllers/TasksController.java:33]: Move this method to `NotificationServiceImpl` Class

File: task-service/src/main/java/com/infodation/task_service/services/TaskServiceImpl.java [hunk 2]
```diff
@@ -29,18 +28,33 @@
...
+        this.taskAssignmentRepository = taskAssignmentRepository;
+    }
+
+    public void assignTaskToUser(TaskAssignmentDTO taskAssignmentDTO) {
+        try {
+            Task task = taskRepository.findById(taskAssignmentDTO.getTaskId())
+                    .orElseThrow(() -> new IllegalArgumentException("Task not found with ID: " + taskAssignmentDTO.getTaskId()));
+            UserTaskAssignment assignment = new UserTaskAssignment();
+            assignment.setUserId(taskAssignmentDTO.getUserId());
+            assignment.setTask(task);
```
Comment (SonNgocTruong) [inline task-service/src/main/java/com/infodation/task_service/services/TaskServiceImpl.java:52]: Using constructor with arguments for shorten code
```suggestion
            UserTaskAssignment assignment = new UserTaskAssignment(taskAssignmentDTO.getUserId(), task);
```

File: userservice/src/main/java/com/infodation/userservice/models/TaskDTO/TaskAssignmentDTO.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class TaskAssignmentDTO {
+    @Valid
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/models/TaskDTO/TaskAssignmentDTO.java:13]: Can you check this @Valid annotation? Is it redundant ?

File: userservice/src/main/java/com/infodation/userservice/models/dto/notification/CreateNotificationDTO.java [hunk 4]
```diff
@@ -11,11 +11,15 @@
 @NoArgsConstructor
 @AllArgsConstructor
 public class CreateNotificationDTO {
+    @Valid
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/models/dto/notification/CreateNotificationDTO.java:14]: Can you check this @Valid annotation? Is it redundant ?

File: task-service/src/main/java/com/infodation/task_service/controllers/TasksController.java [hunk 5]
```diff
@@ -24,6 +26,45 @@ public TasksController(ITaskService taskService) {
...
+    @PostMapping("/assign")
+    public ResponseEntity<String> assignTaskToUser(@RequestBody TaskAssignmentDTO taskAssignmentDTO) {
+        try {
+            log.info("Assigning task {} to user {}", taskAssignmentDTO.getTaskId(), taskAssignmentDTO.getUserId());
+
+            // Save assignment to database
+            taskService.assignTaskToUser(taskAssignmentDTO);
+
+
+            return ResponseEntity.ok("Task assigned successfully.");
```
Comment (SonNgocTruong) [inline task-service/src/main/java/com/infodation/task_service/controllers/TasksController.java:38]: Please try to format your code. I can see redundant empty lines like this

File: task-service/src/main/java/com/infodation/task_service/services/TaskServiceImpl.java [hunk 6]
```diff
@@ -21,26 +20,47 @@
...
+            sendNotificationToRabbitMQ(taskAssignmentDTO);
+            taskAssignmentRepository.save(assignment);
```
Comment (SonNgocTruong) [inline task-service/src/main/java/com/infodation/task_service/services/TaskServiceImpl.java:55]: You need to send a notification to the RabbitMQ only when saving task into database successfully.
```suggestion
            taskAssignmentRepository.save(assignment);
            sendNotificationToRabbitMQ(taskAssignmentDTO);
```

File: task-service/src/main/java/com/infodation/task_service/services/TaskServiceImpl.java [hunk 7]
```diff
@@ -21,26 +20,47 @@
...
+        } catch (Exception e) {
+            log.error("Error occurred while assigning task: ", e);
+            throw new RuntimeException("Error occurred while assigning task", e);
+        }
+    }
+    public void sendNotificationToRabbitMQ(TaskAssignmentDTO message) {
+        rabbitTemplate.convertAndSend("sendNotificationToRabbitMQ", message);
+        log.info("Messages sent: -----> " + message);
```
Comment (SonNgocTruong) [inline task-service/src/main/java/com/infodation/task_service/services/TaskServiceImpl.java:63]:
```suggestion
        log.debug("Messages sent: -----> " + message);
```

File: userservice/src/main/java/com/infodation/userservice/components/TaskAssignmentListener.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        this.userRepository = userRepository;
+    }
+
+    @RabbitListener(queues = "sendNotificationToRabbitMQ")
+    public void handleNotification(TaskAssignmentDTO taskAssignmentDTO) {
+        try {
+            // Query user information from the User table
+            Optional<User> userOptional = userRepository.findById(Long.valueOf(taskAssignmentDTO.getUserId()));
+            User user = userOptional.get();
+            logger.info("Found user: {}", user.getUserId());
```
Comment (SonNgocTruong) [inline userservice/src/main/java/com/infodation/userservice/components/TaskAssignmentListener.java:38]: Feels like you miss checking the presence of User. Please check null of the user

**Review by SonNgocTruong — APPROVED** [review]
LGTM

**Pull request merged**